src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -491,13 +492,32 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
             CancellationTokenSource cts;
             bool disposeCts;
             bool hasTimeout = _timeout != s_infiniteTimeout;
+            TimeoutState timeoutState = null; // Lazy initialization to avoid unnecessary alocation when timeout is not set
+            Timer timer = null;
             if (hasTimeout || cancellationToken.CanBeCanceled)
             {
                 disposeCts = true;
                 cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _pendingRequestsCts.Token);
                 if (hasTimeout)
                 {
-                    cts.CancelAfter(_timeout);
+                    long timeoutMilliseconds = (long)_timeout.TotalMilliseconds;
+                    if (timeoutMilliseconds < -1 || timeoutMilliseconds > int.MaxValue)
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(_timeout));
+                    }
+
+                    timeoutState = new TimeoutState(cts);
+                    timer = new Timer(s =>
+                    {
+                        var state = (TimeoutState)s;
+                        state.IsFired = true;
+                        try
+                        {
+                            state.CancellationSource.Cancel();
+                        }
+                        catch (ObjectDisposedException) // ObjectDisposedException is expected when the timer fires just after the CTS has been disposed
+                        { }
+                    }, timeoutState, timeoutMilliseconds, Threading.Timeout.Infinite);
```

Comment:
There's an overload taking ```TimeSpan``` directly, is there a reason to use the miliseconds instead? Is it because of the ```Threading.Timeout.Infinite```?

Reply:
Commented above. TimeSpan overload is currently being used.